File: kattis/solved/boggle/boggle.java
import java.io.*;
import java.util.HashSet;

public class boggle {
    private static int[] points = {0, 0, 0, 1, 1, 2, 3, 5, 11};
    private static HashSet<String> words;
    private static String[] grid = new String[4];
    private static boolean[][] visited;

    public static void main(String args[]) throws IOException {
        // read in valid words
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int w = Integer.parseInt(in.readLine());
        words = new HashSet<String>(w);
        for (int i = 0; i < w; i++) {
            words.add(in.readLine());
        }

        in.readLine();
        int b = Integer.parseInt(in.readLine());
        for (int i = 0; i < b; i++) {
            for (int j = 0; j < 4; j++) {
                grid[j] = in.readLine();
            }
            solve();
            if (i + 1 < b) {
                in.readLine();
            }
        }
    }

    // calculate and print the three values for each grid
    private static void solve() {
        int score = 0;
        String best = "";
        int found = 0;

        for (String w: words) {
            if (find(w)) {
                int n = w.length();
                score += points[n];
                found++;
                if (n > best.length()) {
                    best = w;
                } else if (n == best.length()) {
                    best = w.compareTo(best) < 0 ? w : best;
                }
            }
        }

        System.out.printf("%d %s %d\n", score, best, found);
    }

    // look for matching start
    private static boolean find(String word) {
        visited = new boolean[4][4];
        boolean works = false;
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                if (grid[i].charAt(j) == word.charAt(0)) {
                    works = works || recur(word, i, j);
                }
                if (works) {
                    return true;
                }
            }
        }
        return false;
    }

    // dfs in grid for word
    private static boolean recur(String word, int r, int c) {
        if (word.length() == 0) {
            return true;
        }

        if (r < 0 || r > 3 || c < 0 || c > 3 || visited[r][c] || word.charAt(0) != grid[r].charAt(c)) {
            return false;
        }

        visited[r][c] = true;
        boolean works = false;
        works = works || recur(word.substring(1), r+1, c);
        works = works || recur(word.substring(1), r-1, c);
        works = works || recur(word.substring(1), r, c+1);
        works = works || recur(word.substring(1), r, c-1);
        works = works || recur(word.substring(1), r-1, c-1);
        works = works || recur(word.substring(1), r-1, c+1);
        works = works || recur(word.substring(1), r+1, c-1);
        works = works || recur(word.substring(1), r+1, c+1);
        visited[r][c] = false;

        return works;
    }
}

File: kattis/solved/boggle/boggle.py
from sys import stdin, stdout

# determine if word is in the boggle grid
def find(word, grid):
    visited = [[False]*4 for _ in range(4)]
    works = False
    for i in range(4):
        for j in range(4):
            if grid[i][j] == word[0]:
                works |= recur(word, grid, visited, i, j)
            if works:
                return True
    return False

# dfs in grid for word
def recur(word, grid, visited, r, c):
    if len(word) == 0:
        return True
    
    if r < 0 or r > 3 or c < 0 or c > 3 or visited[r][c] or word[0] != grid[r][c]:
        return False
    
    visited[r][c] = True
    works = False
    works = works or recur(word[1:], grid, visited, r+1, c)     # down
    works = works or recur(word[1:], grid, visited, r-1, c)     # up
    works = works or recur(word[1:], grid, visited, r, c+1)     # right
    works = works or recur(word[1:], grid, visited, r, c-1)     # left
    works = works or recur(word[1:], grid, visited, r-1, c-1)   # up left
    works = works or recur(word[1:], grid, visited, r-1, c+1)   # up right
    works = works or recur(word[1:], grid, visited, r+1, c-1)   # down left
    works = works or recur(word[1:], grid, visited, r+1, c+1)   # down right
    visited[r][c] = False

    return works

# calculate the three values for each boggle grid
def solve(board):
    score = 0
    best = ''
    found = 0

    for n in words:
        for w in words[n]:
            if find(w, board):
                score += points[n]
                found += 1
                if n > len(best):
                    best = w
                elif n == len(best):
                    best = min(best, w)

    stdout.write('%d %s %d\n' % (score, best, found))

# read in words, organize in dictionary
w = int(stdin.readline())
words = {i:set() for i in range(1, 9)}
for _ in range(w):
    word = stdin.readline().strip()
    words[len(word)].add(word)

# read grid, count found words
points = [0, 0, 0, 1, 1, 2, 3, 5, 11]
stdin.readline()
b = int(stdin.readline())
for i in range(b):
    grid = [stdin.readline().strip() for _ in range(4)]
    solve(grid)
    if i + 1 < b:
        stdin.readline()
stdout.flush()